### PR TITLE
Bug 1826646: shows alert in eventsources form if knative service is not present in namespace

### PR DESCRIPTION
**Pull request by invincibleJai**
**Fixes**: 
https://issues.redhat.com/browse/ODC-3439

**Analysis / Root cause**: 
In Console EventSources form supports only sink to knative Service and in case of no associated service for that form will be never get enabled and user has not much information apart from form field

**Solution Description**: 
Adds an Alert in EventSources form which would be shown if there are no knative service in that namespace

**Gif/Image**:
Apr-22-2020 13-32-47

Screenshot 2020-04-22 at 1 56 09 PM

**Tests**:
Screenshot 2020-04-23 at 1 54 00 PM

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

**Comment by invincibleJai** [comment]
@serenamarie125 PTAL , needed your review

/kind bug

**Comment by invincibleJai** [comment]
@sahil143 updated , PTAL

Screenshot 2020-04-22 at 5 52 33 PM

File: frontend/packages/knative-plugin/src/components/add/EventSource.tsx [hunk 1]
```diff
@@ -100,7 +102,14 @@ const EventSource: React.FC<Props> = ({
       onReset={history.goBack}
       validationSchema={eventSourceValidationSchema}
     >
-      {(props) => <EventSourceForm {...props} namespace={namespace} projects={projects} />}
+      {(props) => (
+        <EventSourceForm
+          {...props}
+          namespace={namespace}
+          projects={projects}
+          showSinkAlert={service?.loaded && !service?.data?.length}
```
Comment (rohitkrai03) [inline frontend/packages/knative-plugin/src/components/add/EventSource.tsx:110]: @invincibleJai Why not render the alert here in this component only? The alert doesn't seem to be related to the form in anyway.

File: frontend/packages/knative-plugin/src/components/add/EventSourceForm.tsx [hunk 2]
```diff
@@ -21,8 +23,14 @@ const EventSourceForm: React.FC<FormikProps<FormikValues> & OwnProps> = ({
...
   namespace,
   projects,
+  showSinkAlert,
 }) => (
   <FlexForm onSubmit={handleSubmit}>
+    {showSinkAlert && (
+      <Alert variant="default" title="Event Source can not be created" isInline>
+        An event source must sink to Knative Service, but no Knative Service exist in this project
+      </Alert>
+    )}
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/components/add/EventSourceForm.tsx:33]: Move to separate component and have it do it's own fetching of the knative service to show the alert.

Right now this change is spread across 3 different components when it should be isolated to a single component.

Also try to use the firehose hook.

Comment (invincibleJai) [reply frontend/packages/knative-plugin/src/components/add/EventSourceForm.tsx]: Have fixed it and refactored code Projects are also fetched with `useK8sWatchResource`

File: frontend/packages/knative-plugin/src/components/add/EventSource.tsx (hunk 1 above)
Comment (invincibleJai) [reply frontend/packages/knative-plugin/src/components/add/EventSource.tsx]: Done

File: frontend/packages/knative-plugin/src/components/add/ShowResourceAlert.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import { K8sResourceKind } from '@console/internal/module/k8s';
+
+interface ShowResourceAlertProps {
+  title: string;
+  resource: FirehoseResource;
+  message?: string;
+}
+
+const ShowResourceAlert: React.FC<ShowResourceAlertProps> = ({ title, message, resource }) => {
+  const [data, loaded, loadError] = useK8sWatchResource<K8sResourceKind[]>(resource);
```
Comment (sahil143) [inline frontend/packages/knative-plugin/src/components/add/ShowResourceAlert.tsx:14]: @invincibleJai this is needed once when the page is loaded. I think a get `useK8sGet` request would do just fine. WDYT?

Comment (invincibleJai) [reply frontend/packages/knative-plugin/src/components/add/ShowResourceAlert.tsx:14]: i thought `watch` will be good as if user creates a kn service in the same namespace, alert will go once it's there with `get` in that scenario even of service get created post load, it'll be showing alert.

let me know

File: frontend/packages/knative-plugin/src/components/add/EventSourcePage.tsx [hunk 4]
```diff
@@ -2,34 +2,36 @@ import * as React from 'react';
...
+import NoKnativeServiceAlert from './NoKnativeServiceAlert';
+import { knativeServingResourcesServices } from '../../utils/get-knative-resources';
 
 type EventSourcePageProps = RouteComponentProps<{ ns?: string }>;
 
 const EventSourcePage: React.FC<EventSourcePageProps> = ({ match, location }) => {
   const namespace = match.params.ns;
   const searchParams = new URLSearchParams(location.search);
-  const resources = [{ kind: ProjectModel.kind, prop: ProjectModel.id, isList: true }];
+  const knServiceResource = { ...knativeServingResourcesServices(namespace)[0], limit: 1 };
```
Comment (christianvogt) [inline frontend/packages/knative-plugin/src/components/add/EventSourcePage.tsx:19]: I think you should still move this into `NoKnativeServiceAlert` and pass the namespace as a prop.
There's no need for the parent to have to deal with passing along a firehose query because then there's no constrain on the resource or limit.
If you just pass the namespace, internal to `NoKnativeServiceAlert` it knows exactly what to look for.
